File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/action/DropGameAction.java
package org.emulinker.kaillera.controller.v086.action;

import com.google.common.flogger.FluentLogger;
import javax.inject.Inject;
import javax.inject.Singleton;
import org.emulinker.kaillera.controller.messaging.MessageFormatException;
import org.emulinker.kaillera.controller.v086.V086Controller.V086ClientHandler;
import org.emulinker.kaillera.controller.v086.protocol.*;
import org.emulinker.kaillera.model.*;
import org.emulinker.kaillera.model.event.*;
import org.emulinker.kaillera.model.exception.DropGameException;

@Singleton
public class DropGameAction
    implements V086Action<PlayerDrop_Request>, V086GameEventHandler<UserDroppedGameEvent> {
  private static final FluentLogger logger = FluentLogger.forEnclosingClass();

  private static final String DESC = "DropGameAction";

  private int actionCount = 0;
  private int handledCount = 0;

  @Inject
  DropGameAction() {}

  @Override
  public int getActionPerformedCount() {
    return actionCount;
  }

  @Override
  public int getHandledEventCount() {
    return handledCount;
  }

  @Override
  public String toString() {
    return DESC;
  }

  @Override
  public void performAction(PlayerDrop_Request message, V086ClientHandler clientHandler)
      throws FatalActionException {
    actionCount++;

    try {
      clientHandler.getUser().dropGame();
    } catch (DropGameException e) {
      logger.atFine().withCause(e).log("Failed to drop game");
    }
  }

  @Override
  public void handleEvent(UserDroppedGameEvent userDroppedEvent, V086ClientHandler clientHandler) {
    handledCount++;

    try {
      KailleraUser user = userDroppedEvent.getUser();
      int playerNumber = userDroppedEvent.getPlayerNumber();
      //			clientHandler.send(PlayerDrop_Notification.create(clientHandler.getNextMessageNumber(),
      // user.getName(), (byte) game.getPlayerNumber(user)));
      if (user.getStealth() == false)
        clientHandler.send(
            PlayerDrop_Notification.create(
                clientHandler.getNextMessageNumber(), user.getName(), (byte) playerNumber));
    } catch (MessageFormatException e) {
      logger.atSevere().withCause(e).log("Failed to contruct PlayerDrop_Notification message");
    }
  }
}

File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/action/DropGameAction.kt
package org.emulinker.kaillera.controller.v086.action

import com.google.common.flogger.FluentLogger
import javax.inject.Inject
import javax.inject.Singleton
import org.emulinker.kaillera.controller.messaging.MessageFormatException
import org.emulinker.kaillera.controller.v086.V086ClientHandler
import org.emulinker.kaillera.controller.v086.protocol.PlayerDrop_Notification
import org.emulinker.kaillera.controller.v086.protocol.PlayerDrop_Request
import org.emulinker.kaillera.model.event.UserDroppedGameEvent
import org.emulinker.kaillera.model.exception.DropGameException

private val logger = FluentLogger.forEnclosingClass()

@Singleton
class DropGameAction @Inject internal constructor() :
    V086Action<PlayerDrop_Request>, V086GameEventHandler<UserDroppedGameEvent> {
  override var actionPerformedCount = 0
    private set
  override var handledEventCount = 0
    private set

  override fun toString(): String {
    return DESC
  }

  @Throws(FatalActionException::class)
  override fun performAction(message: PlayerDrop_Request, clientHandler: V086ClientHandler) {
    actionPerformedCount++
    try {
      clientHandler.user!!.dropGame()
    } catch (e: DropGameException) {
      logger.atFine().withCause(e).log("Failed to drop game")
    }
  }

  override fun handleEvent(event: UserDroppedGameEvent, clientHandler: V086ClientHandler) {
    handledEventCount++
    try {
      val user = event.user
      val playerNumber = event.playerNumber
      //			clientHandler.send(PlayerDrop_Notification.create(clientHandler.getNextMessageNumber(),
      // user.getName(), (byte) game.getPlayerNumber(user)));
      if (!user.stealth)
          clientHandler.send(
              PlayerDrop_Notification(
                  clientHandler.nextMessageNumber, user.name!!, playerNumber.toByte()))
    } catch (e: MessageFormatException) {
      logger.atSevere().withCause(e).log("Failed to construct PlayerDrop_Notification message")
    }
  }

  companion object {
    private const val DESC = "DropGameAction"
  }
}
